pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -187,38 +207,79 @@ func (r *Reconciler) getOrCreateClusterSchedulingPolicySnapshot(ctx context.Cont
-		if revisionLimit <= 0 {
-			err := fmt.Errorf("invalid clusterResourcePlacement %s: invalid revisionHistoryLimit %d", crpKObj, revisionLimit)
-			klog.ErrorS(controller.NewExpectedBehaviorError(err), "Invalid revisionHistoryLimit value and using default value instead", "clusterResourcePlacement", crpKObj)
-			// use the default value instead
-			revisionLimit = fleetv1beta1.RevisionHistoryLimitDefaultValue
+	if len(sortedList.Items)-revisionHistoryLimit > 0 {
+		// We always delete before creating a new snapshot, the snapshot size should never exceed the limit as there is
+		// no finalizer added and object should be deleted immediately.
+		klog.Warningf("The number of clusterSchedulingPolicySnapshots exceeds the revisionHistoryLimit and it should never happen", "clusterResourcePlacement", klog.KObj(crp), "numberOfSnapshots", len(sortedList.Items), "revisionHistoryLimit", revisionHistoryLimit)
+	}
+
+	for i := 0; i <= len(sortedList.Items)-revisionHistoryLimit; i++ { // need to reserve one slot for the new snapshot
+		if err := r.Client.Delete(ctx, &sortedList.Items[i]); err != nil && !errors.IsNotFound(err) {
+			klog.ErrorS(err, "Failed to delete clusterSchedulingPolicySnapshot", "clusterResourcePlacement", klog.KObj(crp), "clusterSchedulingPolicySnapshot", klog.KObj(&sortedList.Items[i]))
+			return controller.NewAPIServerError(false, err)
 		}
 	}
-	if len(sortedList.Items) < int(revisionLimit) {
+	return nil
+}
+
+// deleteRedundantResourceSnapshots handles multiple snapshots in a group.
+func (r *Reconciler) deleteRedundantResourceSnapshots(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement, revisionHistoryLimit int) error {
+	sortedList, err := r.listSortedResourceSnapshots(ctx, crp)
+	if err != nil {
+		return err
+	}
+
+	if len(sortedList.Items) < revisionHistoryLimit {
+		// If the number of existing snapshots is less than the limit no matter how many snapshots in a group, we don't
+		// need to delete any snapshots.
+		// Skip the checking and deleting.
 		return nil
 	}
-	for i := 0; i <= len(sortedList.Items)-int(revisionLimit); i++ {
+
+	crpKObj := klog.KObj(crp)
+	lastGroupIndex := -1
+	groupCounter := 0
+
+	// delete the snapshots from the end as there are could be multiple snapshots in a group in order to keep the latest
+	// snapshots from the end.
+	for i := len(sortedList.Items) - 1; i >= 0; i-- {
+		snapshotKObj := klog.KObj(&sortedList.Items[i])
+		ii, err := parseResourceIndexFromLabel(&sortedList.Items[i])
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse the resource index label", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", snapshotKObj)
+			return controller.NewUnexpectedBehaviorError(err)
+		}
+		if ii != lastGroupIndex {
+			groupCounter++
+			lastGroupIndex = ii
+		}
+		if groupCounter < revisionHistoryLimit { // need to reserve one slot for the new snapshot
+			continue
+		}
 		if err := r.Client.Delete(ctx, &sortedList.Items[i]); err != nil && !errors.IsNotFound(err) {
```

Comment:
Hi Zhiying! Would you mind adding a comment here explaining that if the number of groups is less than revisionHistoryLimit, no groups will be deleted?

Reply:
I just feel that it would make things a little bit more clear if we explicitly mention that the history limit is enforced on groups.